File: Assignment 5/quickcheck-intro/haskell/intro/Intro.hs
module Intro where

import Test.QuickCheck

prop_plusmax :: Int -> Int -> Int -> Property
prop_plusmax k x y = k + (max x y) === max (k + x) (k + y)

prop_int_plus_associative :: Int -> Int -> Int -> Property
prop_int_plus_associative x y z = x + (y + z) === (x + y) + z

prop_reverse :: [Int] -> Property
prop_reverse xs = reverse(reverse xs) === xs

prop_minusmax :: Int -> Int -> Int -> Property
prop_minusmax k x y = k - (max x y) === max (k - x) (k - y)

prop_double_plus_associative :: Double -> Double -> Double -> Property
prop_double_plus_associative x y z = x + (y + z) === (x + y) + z

File: Assignment 5/quickcheck-intro/erlang/intro/intro.erl
-module(intro).

-include_lib("eqc/include/eqc.hrl").

%% The following two lines are super bad style, except during demonstration
-compile(nowarn_export_all).
-compile(export_all).

prop_plusmax() ->
  ?FORALL({K, X, Y}, {int(), int(), int()},
          eqc:equals(K + max(X, Y),
                     max(K + X, K + Y))).

prop_minusmax() ->
  ?FORALL({K, X, Y}, {int(), int(), int()},
          eqc:equals(K - max(X, Y),
                     max(K - X, K - Y))).

prop_reverse() ->
  ?FORALL(XS, list(int()),
         eqc:equals(lists:reverse(lists:reverse(XS)),
                    XS)).

prop_int_plus_associative() ->
  ?FORALL({X, Y, Z}, {int(), int(), int()},
          eqc:equals(X + (Y + Z),
                     (X + Y) + Z)).

prop_real_plus_associative() ->
  ?FORALL({X, Y, Z}, {real(), real(), real()},
          eqc:equals(X + (Y + Z),
                     (X + Y) + Z)).
